arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -194,30 +194,27 @@ typealias Nel<A> = NonEmptyList<A>
  *
  */
 @higherkind
-class NonEmptyList<out A> private constructor(
+class NonEmptyList<out A>(
   val head: A,
-  val tail: List<A>,
-  val all: List<A>
-) : NonEmptyListOf<A>, Iterable<A> by all {
```

Comment:
We should've already supported all extension functions from the std since we extended `Iterable`.

Reply:
Well `AbstractList` has `Iterable` in its hierarchy so it should be more powerful like this.

Reply:
The problem with AbstractList is that exposes operations that are unsafe in Nel like get.
I think this should just implement Iterable structurally so its used in Kotlin loops etc but does not offer unsafe apis like get. Iterable is weaker and still projects all the collections methods from the std lib to terminal ops based on List. 
For the Kotlin docs, we just need structural extensions not even conforming to Iterable if we didn't want the projections which I believe we do.
> As mentioned before, for iterates through anything that provides an iterator, i.e.
> has a member- or extension-function iterator(), whose return type
> has a member- or extension-function next(), and
> has a member- or extension-function hasNext() that returns Boolean.
> All of these three functions need to be marked as operator.

https://kotlinlang.org/docs/reference/control-flow.html#for-loops